spec/models/ability_spec.rb

```diff
@@ -1,437 +1,240 @@
 require 'rails_helper'
 require 'cancan/matchers'
 
-RSpec.describe Ability, type: :model do
-  subject { ability }
-  let(:ability) { Ability.new(user) }
+# Run this file with
+#   $ rspec spec/models/ability_spec.rb -fd
+# to see the output of specs running inside the shared examples [mdv]
 
-  context 'ability' do
-    context 'when a user is connected' do
-      let!(:user) { create(:user) }
-      let!(:team) { create(:team) }
+RSpec.describe Ability, type: :model do
+  subject(:ability) { Ability.new(user) }
+  let(:user){ nil }
+  let(:other_user) { create(:user) }
+
+  describe "Guest User" do
+    it_behaves_like "can view public pages"
+    it_behaves_like "can not modify things on public pages"
+    it_behaves_like "can not create new user"
+    it_behaves_like "can not comment"
+    it_behaves_like "has no access to other user's accounts"
+    it_behaves_like "can not read role restricted or owner restricted pages"
+    it "does not have an account" do
+      true
+    end
+  end
 
-      describe 'she/he is allowed to do everything on her/his account' do
-        it { expect(ability).to be_able_to(:show, user) }
-        it { expect(ability).not_to be_able_to(:create, User.new) } # this only happens through GitHub
+  describe "Logged in user, not confirmed" do
+    let(:user) { build_stubbed(:user, :unconfirmed) }
 
-        it { expect(ability).to be_able_to(:resend_confirmation_instruction, user) }
-      end
+    it_behaves_like 'same as guest user'
+    it_behaves_like "can modify own account"
+  end
 
-      context 'when a user is admin' do
-        let(:organizer_role) { create(:organizer_role, user: user) }
-        it "should be able to CRUD on anyone's account" do
-          expect(subject).to be_able_to(:crud, organizer_role)
-        end
-      end
+  describe "Confirmed user" do
+    let!(:user) { create(:user) }
 
-      describe 'she/he is not allowed to CRUD on someone else account' do
-        let(:other_user) { create(:user) }
-        it { expect(ability).not_to be_able_to(:show, other_user) }
+    it_behaves_like "same as logged in user"
+    it_behaves_like "can create a Project"
+    it_behaves_like "can see mailings list too"
+    it_behaves_like "can read mailings sent to them"
+    it_behaves_like "can comment now" # not implemented; false positives; needs work
+    it { expect(subject).to be_able_to([:join, :create], Team) }
```

Comment:
Cosmetic, but it's not necessary to write `subject` when using the short hand syntax. You can simply write:
```ruby
it { is_expected.to be_able_to %i(join create), Team }
```

Same for all further appearance 😉